src/engine/interop/specification.md

```diff
@@ -0,0 +1,151 @@
+# Engine API. Interop edition
+
+This document specifies a subset of the Engine API methods that are required to be implemented for the Merge interop.
+
+*Note*: Sync API design is yet a draft and considered optional for the interop.
+
+## Underlying protocol
+
+This specification is based on [Ethereum JSON-RPC API](https://eth.wiki/json-rpc/API) and inherits the following properties of this standard:
+* Supported communication protocols (HTTP and WebSocket)
+* Message format and encoding notation
+* [Error codes improvement proposal](https://eth.wiki/json-rpc/json-rpc-error-codes-improvement-proposal)
+
+Client software **MUST** expose Engine API at a port independent from JSON-RPC API. The default port for the Engine API is 8550 for HTTP and 8551 for WebSocket.
+
+## Error codes
+
+The list of error codes introduced by this specification can be found below.
+
+| Code | Possible Return message | Description |
+| - | - | - |
+| 4 | Unknown block | Should be used when a call refers to the unknown block |
+| 5 | Unavailable payload | Should be used when the `payloadId` parameter of `engine_getPayload` call refers to the payload building process that is unavailable |
+
+## Structures
+
+Fields having `DATA` and `QUANTITY` types **MUST** be encoded according to the [HEX value encoding](https://eth.wiki/json-rpc/API#hex-value-encoding) section of Ethereum JSON-RPC API.
+
+*Note:* Byte order of encoded value having `QUANTITY` type is big-endian.
+
+### ExecutionPayload
+
+This structure maps on the [`ExecutionPayload`](https://github.com/ethereum/consensus-specs/blob/dev/specs/merge/beacon-chain.md#ExecutionPayload) structure of the beacon chain spec. The fields are encoded as follows:
+- `parentHash`: `DATA`, 32 Bytes
+- `coinbase`:  `DATA`, 20 Bytes
+- `stateRoot`: `DATA`, 32 Bytes
+- `receiptRoot`: `DATA`, 32 Bytes
+- `logsBloom`: `DATA`, 256 Bytes
+- `random`: `DATA`, 32 Bytes
+- `blockNumber`: `QUANTITY`, 64 Bits
+- `gasLimit`: `QUANTITY`, 64 Bits
+- `gasUsed`: `QUANTITY`, 64 Bits
+- `timestamp`: `QUANTITY`, 64 Bits
+- `extraData`: `DATA`, 0 to 32 Bytes
+- `baseFeePerGas`: `QUANTITY`, 256 Bits
+- `blockHash`: `DATA`, 32 Bytes
+- `transactions`: `Array of DATA` - Array of transaction objects, each object is a byte list (`DATA`) representing `TransactionType || TransactionPayload` or `LegacyTransaction` as defined in [EIP-2718](https://eips.ethereum.org/EIPS/eip-2718)
+
+## Core
+
+### engine_preparePayload
+
+#### Parameters
+1. `Object` - The payload attributes:
+- `parentHash`: `DATA`, 32 Bytes - hash of the parent block
+- `timestamp`: `QUANTITY`, 64 Bits - value for the `timestamp` field of the new payload
```

Comment:
Unrelated to the api - is there a concern with the beacon node asking the execution engine to prepare a payload with a timestamp that would be considered invalid under old consensus rules?

Reply:
The `timestamp` field of the payload is set to the `slot.timestamp` and this condition is enforced by validation conditions on the beacon chain, so the timestamp value is enforced to be aligned with the timestamp of a slot. Beacon chain has strict time restrictions and validations in the fork choice and gossip. So, I think that execution clients should blindly follow the consensus clients with respect to the timestamp validation, i.e. don't validate timestamps at all to avoid edge case bugs.

Reply:
This is a good point, btw! Thanks for raising it 👍